geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/GroupBy.scala

```diff
@@ -0,0 +1,113 @@
+/***********************************************************************
+* Copyright (c) 2013-2016 Commonwealth Computer Research, Inc.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Apache License, Version 2.0
+* which accompanies this distribution and is available at
+* http://www.opensource.org/licenses/apache2.0.php.
+*************************************************************************/
+
+package org.locationtech.geomesa.utils.stats
+
+import org.opengis.feature.simple.SimpleFeature
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+case class GroupBy[T](attribute: Int, exampleStat: Stat)(implicit ct: ClassTag[T]) extends Stat {
+  override type S = GroupBy[T]
+
+  val groupedStats: mutable.HashMap[T, Stat] = mutable.HashMap[T, Stat]()
+
+  /**
+    * Compute statistics based upon the given simple feature.
+    * This method will be called for every SimpleFeature a query returns.
+    *
+    * @param sf feature to evaluate
+    */
+  override def observe(sf: SimpleFeature): Unit = {
+    val key = sf.getAttribute(attribute).asInstanceOf[T]
+    groupedStats.get(key) match {
```

Comment:
can simplify this: `groupedStats.getOrElseUpdate(key, exampleStat.newcopy).observe(sf)`